src/supernova/snark.rs

```diff
@@ -0,0 +1,656 @@
+    }
+  }
+
+  #[derive(Clone)]
+  struct CubeCircuit<G: Group> {
+    _p: PhantomData<G>,
+  }
+
+  impl<G: Group> StepCircuit<G::Scalar> for CubeCircuit<G> {
+    fn arity(&self) -> usize {
+      1
+    }
+
+    fn circuit_index(&self) -> usize {
+      1
+    }
+
+    fn synthesize<CS: ConstraintSystem<G::Scalar>>(
+      &self,
+      cs: &mut CS,
+      _pc: Option<&AllocatedNum<G::Scalar>>,
+      z: &[AllocatedNum<G::Scalar>],
+    ) -> Result<
+      (
+        Option<AllocatedNum<G::Scalar>>,
+        Vec<AllocatedNum<G::Scalar>>,
+      ),
+      SynthesisError,
+    > {
+      let z_i = &z[0];
+
+      let z_sq = z_i.square(cs.namespace(|| "z_i^2"))?;
+      let z_cu = z_sq.mul(cs.namespace(|| "z_i^3"), z_i)?;
+
+      let next_pc = AllocatedNum::alloc(cs.namespace(|| "next_pc"), || Ok(G::Scalar::from(0u64)))?;
+
+      cs.enforce(
+        || "next_pc = 0",
+        |lc| lc + CS::one(),
+        |lc| lc + next_pc.get_variable(),
+        |lc| lc,
+      );
+
+      Ok((Some(next_pc), vec![z_cu]))
+    }
+  }
+
+  #[derive(Clone)]
+  enum TestCircuit<G: Group> {
+    Square(SquareCircuit<G>),
+    Cube(CubeCircuit<G>),
+  }
+
+  impl<G: Group> StepCircuit<G::Scalar> for TestCircuit<G> {
+    fn arity(&self) -> usize {
+      1
+    }
+
+    fn circuit_index(&self) -> usize {
+      match self {
+        TestCircuit::Square(c) => c.circuit_index(),
+        TestCircuit::Cube(c) => c.circuit_index(),
+      }
+    }
+
+    fn synthesize<CS: ConstraintSystem<G::Scalar>>(
+      &self,
+      cs: &mut CS,
+      pc: Option<&AllocatedNum<G::Scalar>>,
+      z: &[AllocatedNum<G::Scalar>],
+    ) -> Result<
+      (
+        Option<AllocatedNum<G::Scalar>>,
+        Vec<AllocatedNum<G::Scalar>>,
+      ),
+      SynthesisError,
+    > {
+      match self {
+        TestCircuit::Square(c) => c.synthesize(cs, pc, z),
+        TestCircuit::Cube(c) => c.synthesize(cs, pc, z),
+      }
+    }
+  }
+
+  struct TestNIVC<G1, G2> {
+    _p: PhantomData<(G1, G2)>,
+  }
+
+  impl<G1, G2> TestNIVC<G1, G2> {
+    fn new() -> Self {
+      TestNIVC { _p: PhantomData }
+    }
+  }
+
+  impl<G1, G2> NonUniformCircuit<G1, G2, TestCircuit<G1>, TrivialSecondaryCircuit<G2::Scalar>>
+    for TestNIVC<G1, G2>
+  where
+    G1: Group<Base = <G2 as Group>::Scalar>,
+    G2: Group<Base = <G1 as Group>::Scalar>,
+  {
+    fn initial_program_counter(&self) -> <G1 as Group>::Scalar {
+      G1::Scalar::from(0u64)
+    }
+
+    fn num_circuits(&self) -> usize {
+      2
+    }
+
+    fn primary_circuit(&self, circuit_index: usize) -> TestCircuit<G1> {
+      match circuit_index {
+        0 => TestCircuit::Square(SquareCircuit { _p: PhantomData }),
+        1 => TestCircuit::Cube(CubeCircuit { _p: PhantomData }),
+        _ => panic!("Invalid circuit index"),
+      }
+    }
+
+    fn secondary_circuit(&self) -> TrivialSecondaryCircuit<G2::Scalar> {
+      Default::default()
+    }
+  }
+
+  fn test_nivc_trivial_with_compression_with<G1, G2, E1, E2>()
```

Comment:
We should make sure to add `CompressedSNARK` tests when more NIVC tests become available (#52).